classes/TravelManager.php

```diff
@@ -0,0 +1,261 @@
+        $this->user->updateData();
+        return true;
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function enterPortal($portal_id): bool {
+        $ignore_travel_restrictions = $this->user->isHeadAdmin();
+        // portal data
+        $portal_data = Travel::getPortalData($this->system, $portal_id);
+        if (empty($portal_data)) {
+            throw new Exception('You cannot enter here!');
+        }
+        if (!$this->checkRestrictions()) {
+            throw new Exception('Unable to move!');
+        }
+        // check if the player is at the correct entrance
+        if (($this->user->location->z !== $portal_data['from_id']
+            || $this->user->location->x !== $portal_data['entrance_x']
+            || $this->user->location->y !== $portal_data['entrance_y'])
+            && !$ignore_travel_restrictions) {
+            throw new Exception('You cannot enter here!');
+        }
+        // check if the player is in a faction that allows this portal
+        $portal_whitelist = array_map('trim', explode(',', $portal_data['whitelist']));
+        if (!in_array($this->user->village, $portal_whitelist) && !$ignore_travel_restrictions) {
+            throw new Exception('You are unable to enter here!');
+        }
+        // update the player data
+        $this->user->location->x = $portal_data['exit_x'];
+        $this->user->location->y = $portal_data['exit_y'];
+        $this->user->location->z = $portal_data['to_id'];
+        $this->user->last_movement_ms = System::currentTimeMs();
+        $this->user->updateData();
+        return true;
+    }
+
+    public function fetchScoutData(): array {
+        return $this->fetchNearbyPlayers();
+    }
+
+    public function fetchNearbyPlayers(): array {
+        $sql = "SELECT `users`.`user_id`, `users`.`user_name`, `users`.`village`, `users`.`rank`, `users`.`stealth`,
+                `users`.`level`, `users`.`attack_id`, `users`.`battle_id`, `ranks`.`name`, `users`.`location`
+                FROM `users`
+                INNER JOIN `ranks`
+                ON `users`.`rank`=`ranks`.`rank_id`
+                WHERE `users`.`last_active` > UNIX_TIMESTAMP() - 120
+                ORDER BY `users`.`exp` DESC";
+        $result = $this->system->query($sql);
+        $users = $this->system->db_fetch_all($result);
+        $return_arr = [];
+        foreach ($users as $user) {
+            // check if the user is nearby (including stealth
+            $scout_range = max(0, $this->user->scout_range - $user['stealth']);
+            $location = new TravelCoords($user['location']);
+            if ($location->z !== $this->user->location->z ||
+                abs($location->x - $this->user->location->x) > $scout_range ||
+                abs($location->y - $this->user->location->y) > $scout_range) {
+                continue;
+            }
+            $user['location_array'] = [$location->x, $location->y, $location->z];
```

Comment:
Generally a fan of keeping this structured so we don't get x/y mixed up as y/x by mistake, could we make this explicitly `$user['x']`, `$user['y']`, `$user['map_id']`?